en/docs/extended_topics/user_op.md

```diff
@@ -1,414 +1,816 @@
+      int64_t num_axes = ctx->LogicalTensorDesc4InputArgNameAndIndex("in", 0).shape().NumAxes();
+      for (int64_t i = 0; i < num_axes; ++i) {
+        ctx->NewBuilder().Split(ctx->inputs(), i).Split(user_op::OpArg("out", 0), i).Build();
+      }
+      ctx->NewBuilder().PartialSum(ctx->inputs()).PartialSum(user_op::OpArg("out", 0)).Build();
+      return Maybe<void>::Ok();
+    });
 ```
 
+## Detailed Introduction of OpKernelRegistry
+
+### SetInferTmpSizeFn
 
+In some kernel implementations of op, some extra buffer may be required to store temporary data during the `Compute`.
 
-#### DataType Infer Function
+We can specify the buffer size when registering the kernel by using the `SetInferTmpSizeFn`. Then get the buffer and use it in the `Compute` function.
 
-多数Op的input tensor和output tensor的类型相同，但对于一些特殊Op（如`Cast` op）来说，其需要传入一个data_type infer function来推导output tensor的类型。
+The following code registers the kernel with `SetInferTmpSizeFn` to specify a buffer size as 1024 bytes:
 
 ``` cpp
-.SetDataTypeInferFn([](user_op::InferContext* ctx) {
-      DataType* out_tensor_type = ctx->Dtype4ArgNameAndIndex("out", 0);
-      *out_tensor_type = DataType::kDouble;
-      return Maybe<void>::Ok();
-    })
+REGISTER_USER_KERNEL("XOp")
+    .SetInferTmpSizeFn(
+      [](const oneflow::user_op::InferContext*) {
+         return 1024; 
+      });
 ```
 
-上述代码就是把out tensor的数据类型设置为`double`。
+Once the buffer size is set by `SetInferTmpSizeFn`, this buffer can be retrieved in `Compute` by calling the `KernelContext::Sensor4ArgNameAndIndex`. This buffer is encapsulated as `oneflow::user_op::Sensor` which can be converted to other types of pointers by calling the `dptr` or `mut_dptr`.
 
-对于无需更改数据类型的Op，也无需在注册Op时指定`SetDataTypeInferFn()`，因为OneFlow框架提供的默认实现就是让output tensors 和 input tensors 的数据类型一致。
+```cpp
+class XKernel final : public oneflow::user_op::OpKernel {
+  void Compute(oneflow::user_op::KernelContext* ctx) override {
+    oneflow::user_op::Tensor* tmp = ctx->Tensor4ArgNameAndIndex("tmp_buffer", 0);
+
+    //The conversion yields a char* buffer of 1024 bytes.
+    char* pBuff = tmp->mut_dptr<char>();
+
+    ...
+  }
+};
+```
 
+## Detailed Introduction of OpGradRegistry
 
+Oneflow is automatically get derivative during backward map expansion and the OneFlow framework uses [Automatic Differentiation](https://en.wikipedia.org/wiki/Automatic_differentiation) to get the derivative which means automatically find the gradient of the entire expression using the chain rule.
 
-#### Sbp Function、BatchAxis
+In order to automatically get derivative a custom op, we need to register it with `REGISTER_USER_OP_GRAD`. From a mathematical point of view, the registration process is the computation of the backward derivation that we specify for our custom op. From a programming point of view, it is to set up a backward-generating function for a custom op. Within that function, write code that specifies how the input gradient of that op is to be calculated.
 
```

Comment:
we write code that

Reply:
In order to automatically get derivative a custom op -> In order to get gradient automatically

Reply:
the registration process is the computation of the backward derivation that we specify for our custom op -> the registration process is that we specify the calculation method of backward derivation for the custom Op